cpp/foxglove-websocket/include/foxglove/websocket/websocket_server.hpp

```diff
@@ -802,98 +895,135 @@ inline void Server<ServerConfiguration>::handleTextMessage(ConnHandle hdl, const
       sendJsonRaw(hdl, jsonMsg.dump());
     } break;
     case UNSUBSCRIBE_CONNECTION_GRAPH: {
+      if (!_handlers.subscribeConnectionGraphHandler) {
+        return;
+      }
+
       if (clientInfo.subscribedToConnectionGraph) {
         clientInfo.subscribedToConnectionGraph = false;
-        std::unique_lock<std::shared_mutex> lock(_connectionGraphMutex);
-        _connectionGraph.subscriptionCount--;
-        if (_connectionGraph.subscriptionCount == 0 && _handlers.subscribeConnectionGraphHandler) {
+        bool unsubscribeFromConnnectionGraph = false;
+        {
+          std::unique_lock<std::shared_mutex> lock(_connectionGraphMutex);
+          _connectionGraph.subscriptionCount--;
+          unsubscribeFromConnnectionGraph = _connectionGraph.subscriptionCount == 0;
+        }
+        if (unsubscribeFromConnnectionGraph) {
           _server.get_alog().write(APP, "Unsubscribing from connection graph updates.");
           _handlers.subscribeConnectionGraphHandler(false);
         }
       } else {
-        sendStatus(hdl, StatusLevel::Error,
-                   "Client was not subscribed to connection graph updates");
+        sendStatusAndLogMsg(hdl, StatusLevel::Error,
+                            "Client was not subscribed to connection graph updates");
+      }
+    } break;
+    case FETCH_ASSET: {
+      if (!_handlers.fetchAssetHandler) {
+        return;
+      }
+
+      const auto uri = payload.at("uri").get<std::string>();
+      const auto requestId = payload.at("requestId").get<uint32_t>();
+
+      try {
+        _handlers.fetchAssetHandler(uri, requestId, hdl);
+      } catch (const std::exception& e) {
+        sendStatusAndLogMsg(hdl, StatusLevel::Error, e.what());
+      } catch (...) {
+        sendStatusAndLogMsg(hdl, StatusLevel::Error, op + ": Failed to execute handler");
```

Comment:
If you want to avoid repeating these 6-7 lines every time we call a handler, it could be abstracted into the handlers struct or a helper method to call the handler and log if it failed.